diff --git a/runorraise.c b/runorraise.c
new file mode 100644
index 0000000..5494aaa
--- /dev/null
+++ b/runorraise.c
@@ -0,0 +1,52 @@
+typedef struct {
+	char *class;
+	char *name;
+	char *spawncmd[];
+} ClassHintedCmd;
+
+void
+runorraise(const Arg *_arg) {
+	ClassHintedCmd *arg = *(ClassHintedCmd **)_arg;
+	Arg a = { .ui = ~0 };
+	Monitor *mon;
+	Client *c;
+	XClassHint hint = { NULL, NULL };
+
+	if (arg->class == NULL && arg->name == NULL) {
+		return;
+	}
+
+	/* Tries to find the client */
+	for (mon = mons; mon; mon = mon->next) {
+		for (c = mon->clients; c; c = c->next) {
+			XGetClassHint(dpy, c->win, &hint);
+
+			if (arg->class && arg->name
+				&& strcmp(arg->class, hint.res_class) != 0
+				|| strcmp(arg->name, hint.res_name) != 0) {
+				continue;
+
+			} else {
+
+				} if (arg->class && hint.res_class
+					&& strcmp(arg->class, hint.res_class) != 0) {
+					continue;
+
+				if (arg->name && hint.res_name
+					&& strcmp(arg->name, hint.res_name) != 0) {
+					continue;
+				}
+			}
+
+			a.ui = c->tags;
+			view(&a);
+			focus(c);
+			XRaiseWindow(dpy, c->win);
+			return;
+
+		}
+	}
+	/* Client not found: spawn it */
+	spawn(&(Arg){.v = arg->spawncmd});
+}
+