From 2198195e514ef54153281fb1ec062425be2b3aa7 Mon Sep 17 00:00:00 2001
From: Jonathen Russell <jgr006@gmail.com>
Date: Mon, 8 Jan 2018 12:19:54 +1000
Subject: [PATCH] Apply save floats patch

---
 dwm.c | 42 +++++++++++++++++++++++++++++++++++++++---
 1 file changed, 39 insertions(+), 3 deletions(-)

diff --git a/dwm.c b/dwm.c
index ec6a27c..fa88776 100644
--- a/dwm.c
+++ b/dwm.c
@@ -89,6 +89,7 @@ struct Client {
 	char name[256];
 	float mina, maxa;
 	int x, y, w, h;
+	int sfx, sfy, sfw, sfh; /* stored float geometry, used on mode revert */
 	int oldx, oldy, oldw, oldh;
 	int basew, baseh, incw, inch, maxw, maxh, minw, minh;
 	int bw, oldbw;
@@ -1057,6 +1058,10 @@ manage(Window w, XWindowAttributes *wa)
 	updatewindowtype(c);
 	updatesizehints(c);
 	updatewmhints(c);
+	c->sfx = c->x;
+	c->sfy = c->y;
+	c->sfw = c->w;
+	c->sfh = c->h;
 	XSelectInput(dpy, w, EnterWindowMask|FocusChangeMask|PropertyChangeMask|StructureNotifyMask);
 	grabbuttons(c, 0);
 	if (!c->isfloating)
@@ -1501,10 +1506,33 @@ setfullscreen(Client *c, int fullscreen)
 void
 setlayout(const Arg *arg)
 {
+	Monitor *m;
+	Client *c;
+
+	unsigned int fromlt = selmon->sellt;
+
 	if (!arg || !arg->v || arg->v != selmon->lt[selmon->sellt])
 		selmon->sellt ^= 1;
 	if (arg && arg->v)
 		selmon->lt[selmon->sellt] = (Layout *)arg->v;
+
+	/* switch from floating */
+	if (selmon->lt[fromlt]->arrange == NULL)
+		/*save last known float dimensions*/
+		for (m = mons; m; m = m->next)
+			for (c = m->clients; c; c = c->next) {
+				c->sfx = c->x;
+				c->sfy = c->y;
+				c->sfw = c->w;
+				c->sfh = c->h;
+			}
+	/* switch to floating */
+	else if (selmon->lt[selmon->sellt]->arrange == NULL)
+		/*restore last known float dimensions*/
+		for (m = mons; m; m = m->next)
+			for (c = m->clients; c; c = c->next)
+				resize(c, c->sfx, c->sfy, c->sfw, c->sfh, 0);
+
 	strncpy(selmon->ltsymbol, selmon->lt[selmon->sellt]->symbol, sizeof selmon->ltsymbol);
 	if (selmon->sel)
 		arrange(selmon);
@@ -1714,9 +1742,17 @@ togglefloating(const Arg *arg)
 	if (selmon->sel->isfullscreen) /* no support for fullscreen windows */
 		return;
 	selmon->sel->isfloating = !selmon->sel->isfloating || selmon->sel->isfixed;
-	if (selmon->sel->isfloating)
-		resize(selmon->sel, selmon->sel->x, selmon->sel->y,
-			selmon->sel->w, selmon->sel->h, 0);
+	if (selmon->sel->isfloating) {
+		/*restore last known float dimensions*/
+		resize(selmon->sel, selmon->sel->sfx, selmon->sel->sfy,
+		       selmon->sel->sfw, selmon->sel->sfh, 0);
+	} else {
+		/*save last known float dimensions*/
+		selmon->sel->sfx = selmon->sel->x;
+		selmon->sel->sfy = selmon->sel->y;
+		selmon->sel->sfw = selmon->sel->w;
+		selmon->sel->sfh = selmon->sel->h;
+	}
 	arrange(selmon);
 }
 
-- 
2.15.1

